Lineage finding issue

It looks like fetching lineage for files from the https://dev.renku.ch/projects/tasko.olevski/nccr-demo-test-1 project fails.

https://dev.renku.ch/projects/tasko.olevski/nccr-demo-test-1/files/lineage/data/evals/random_forest_train_set_perf.png

```
2023-04-24 09:32:17,385 ERROR application - Lineage generation failed
java.lang.Exception: Finding lineage for 'tasko.olevski/nccr-demo-test-1' and 'data/evals/random_forest_train_set_perf.png' failed
	at io.renku.knowledgegraph.projects.files.lineage.LineageFinderImpl$$anonfun$loggingError$1.applyOrElse(LineageFinder.scala:85)
	at io.renku.knowledgegraph.projects.files.lineage.LineageFinderImpl$$anonfun$loggingError$1.applyOrElse(LineageFinder.scala:81)
	at cats.ApplicativeError.$anonfun$recoverWith$1(ApplicativeError.scala:172)
	at main$ @ io.renku.knowledgegraph.Microservice$.main(Microservice.scala:35)
	at parSequence$extension @ io.renku.knowledgegraph.users.projects.finder.GLProjectFinderImpl.fetchAllOtherPagesPar(GLProjectFinder.scala:68)
	at sequence @ io.renku.knowledgegraph.metrics.KGMetricsImpl.$anonfun$updateCounts$2(KGMetrics.scala:48)
	at sequence @ io.renku.knowledgegraph.metrics.KGMetricsImpl.$anonfun$updateCounts$2(KGMetrics.scala:48)
	at map @ io.renku.knowledgegraph.projects.files.lineage.NodeDetailsFinderImpl.findDetails(NodeDetailsFinder.scala:61)
Caused by: java.lang.IllegalArgumentException: No plan with https://dev.renku.ch/activities/3a9276d39788403b96bf378e74fa1546
at io.renku.knowledgegraph.projects.files.lineage.NodeDetailsFinderImpl.$anonfun$failIf$1(NodeDetailsFinder.scala:105)
	at >>$extension @ fs2.Pull$.fs2$Pull$$go$1(Pull.scala:1240)
	at foreverM$extension @ io.renku.knowledgegraph.metrics.KGMetricsImpl.run(KGMetrics.scala:42)
	at >>$extension @ fs2.Pull$.fs2$Pull$$go$1(Pull.scala:1240)
	at foreverM$extension @ io.renku.knowledgegraph.metrics.KGMetricsImpl.run(KGMetrics.scala:42)
	at delay @ fs2.io.net.SocketGroupCompanionPlatform$AsyncSocketGroup.$anonfun$client$2(SocketGroupPlatform.scala:55)
	at guarantee$extension @ fs2.internal.InterruptContext.complete(InterruptContext.scala:49)
	at >>$extension @ fs2.Pull$.fs2$Pull$$go$1(Pull.scala:1240)
	at foreverM$extension @ io.renku.knowledgegraph.metrics.KGMetricsImpl.run(KGMetrics.scala:42)
	at delay @ fs2.io.net.SocketCompanionPlatform.$anonfun$forAsync$2(SocketPlatform.scala:43)
	at guarantee$extension @ fs2.internal.InterruptContext.complete(InterruptContext.scala:49)
	at delay @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.endOfOutput(SocketPlatform.scala:167)

--- a/knowledge-graph/src/main/scala/io/renku/knowledgegraph/projects/files/lineage/NodeDetailsFinder.scala
+++ b/knowledge-graph/src/main/scala/io/renku/knowledgegraph/projects/files/lineage/NodeDetailsFinder.scala
@@ -24,13 +24,14 @@ import cats.syntax.all._
 import eu.timepit.refined.auto._
 import io.circe.{Decoder, DecodingFailure}
 import io.renku.graph.config.RenkuUrlLoader
+import io.renku.graph.model.{projects, GraphClass, RenkuUrl}
 import io.renku.graph.model.Schemas._
 import io.renku.graph.model.projects.ResourceId
-import io.renku.graph.model.views.RdfResource
-import io.renku.graph.model.{GraphClass, RenkuUrl, projects}
+import io.renku.jsonld.syntax._
 import io.renku.tinytypes.json.TinyTypeDecoders
-import io.renku.triplesstore.SparqlQuery.Prefixes
 import io.renku.triplesstore._
+import io.renku.triplesstore.SparqlQuery.Prefixes
+import io.renku.triplesstore.client.syntax._
 import model.{ExecutionInfo, Node}
 import org.typelevel.log4cats.Logger
 
@@ -102,7 +103,7 @@ private class NodeDetailsFinderImpl[F[_]: Async: Parallel: Logger: SparqlQueryTi
         case location: Node.Location =>
           new IllegalArgumentException(s"No entity with $location").raiseError[F, Node]
         case runInfo: ExecutionInfo =>
-          new IllegalArgumentException(s"No plan with ${runInfo.entityId}").raiseError[F, Node]
+          new IllegalArgumentException(s"No activity with ${runInfo.entityId}").raiseError[F, Node]
         case other =>
           new IllegalArgumentException(s"Entity $other not recognisable").raiseError[F, Node]
       }
@@ -120,29 +121,29 @@ private object NodeDetailsFinder {
     SparqlQuery.of(
       name = "lineage - entity details",
       Prefixes of (prov -> "prov", schema -> "schema", renku -> "renku"),
-      s"""|SELECT DISTINCT ?type ?location ?label
-          |WHERE {
-          |  {
-          |    GRAPH <${GraphClass.Project.id(projectId)}> {
-          |      ?entity prov:atLocation '$location';
-          |              a prov:Entity;
-          |              a ?type
-          |    }
-          |  } {
-          |    GRAPH <${GraphClass.Project.id(projectId)}> {
-          |      ?activityId a prov:Activity;
-          |                  prov:qualifiedUsage / prov:entity ?entity;
-          |                  ^renku:hasActivity ${projectId.showAs[RdfResource]}
-          |    }
-          |  } UNION {
-          |    GRAPH <${GraphClass.Project.id(projectId)}> {
-          |      ?entity prov:qualifiedGeneration / prov:activity / ^renku:hasActivity ${projectId.showAs[RdfResource]}
-          |    }
-          |  }
-          |  BIND ('$location' AS ?location)
-          |  BIND ('$location' AS ?label)
-          |}
-          |""".stripMargin
+      sparql"""|SELECT DISTINCT ?type ?location ?label
+               |WHERE {
+               |  {
+               |    GRAPH ${GraphClass.Project.id(projectId)} {
+               |      ?entity prov:atLocation ${location.asObject};
+               |              a prov:Entity;
+               |              a ?type
+               |    }
+               |  } {
+               |    GRAPH ${GraphClass.Project.id(projectId)} {
+               |      ?activityId a prov:Activity;
+               |                  prov:qualifiedUsage / prov:entity ?entity;
+               |                  ^renku:hasActivity ${projectId.asEntityId}
+               |    }
+               |  } UNION {
+               |    GRAPH ${GraphClass.Project.id(projectId)} {
+               |      ?entity prov:qualifiedGeneration / prov:activity / ^renku:hasActivity ${projectId.asEntityId}
+               |    }
+               |  }
+               |  BIND (${location.asObject} AS ?location)
+               |  BIND (${location.asObject} AS ?label)
+               |}
+               |""".stripMargin
     )
 
   implicit val activityIdQuery: (ExecutionInfo, ResourceId) => SparqlQuery = {
@@ -150,73 +151,73 @@ private object NodeDetailsFinder {
       SparqlQuery.of(
         name = "lineage - plan details",
         Prefixes of (prov -> "prov", renku -> "renku", schema -> "schema"),
-        s"""|SELECT DISTINCT ?type (CONCAT(STR(?command), (GROUP_CONCAT(?commandParameter; separator=' '))) AS ?label) ?location
-            |WHERE {
-            |  {
-            |    SELECT DISTINCT ?command ?type ?location
-            |    WHERE {
-            |      GRAPH <${GraphClass.Project.id(projectId)}> {
-            |        <$activityId> prov:qualifiedAssociation/prov:hadPlan ?planId;
-            |                      a ?type.
-            |        OPTIONAL { ?planId renku:command ?maybeCommand. }
-            |      }
-            |      BIND (IF(bound(?maybeCommand), CONCAT(STR(?maybeCommand), STR(' ')), '') AS ?command).
-            |      BIND (<$activityId> AS ?location)
-            |    }
-            |  } {
-            |    SELECT ?position ?commandParameter
-            |     WHERE {
-            |      { # inputs
-            |        GRAPH <${GraphClass.Project.id(projectId)}> {
-            |          <$activityId> prov:qualifiedAssociation/prov:hadPlan ?planId.
-            |          ?planId renku:hasInputs ?input .
-            |          ?paramValue a renku:ParameterValue ;
-            |                      schema:valueReference ?input .
-            |          ?paramValue schema:value ?value .
-            |          OPTIONAL { ?input renku:mappedTo/renku:streamType ?maybeStreamType. }
-            |          ?input renku:position ?position.
-            |          OPTIONAL { ?input renku:prefix ?maybePrefix }
-            |        }
-            |        BIND (IF(bound(?maybeStreamType), ?maybeStreamType, '') AS ?streamType).
-            |        BIND (IF(?streamType = 'stdin', '< ', '') AS ?streamOperator).
-            |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix).
-            |        BIND (CONCAT(?prefix, ?streamOperator, STR(?value)) AS ?commandParameter) .
-            |      } UNION { # outputs
-            |        GRAPH <${GraphClass.Project.id(projectId)}> {
-            |          <$activityId> prov:qualifiedAssociation/prov:hadPlan ?planId.
-            |          ?planId renku:hasOutputs ?output .
-            |          ?paramValue a renku:ParameterValue ;
-            |                      schema:valueReference ?output .
-            |          ?paramValue schema:value ?value .
-            |          ?output renku:position ?position.
-            |          OPTIONAL { ?output renku:mappedTo/renku:streamType ?maybeStreamType. }
-            |          OPTIONAL { ?output renku:prefix ?maybePrefix }
-            |        }
-            |        BIND (IF(bound(?maybeStreamType), ?maybeStreamType, '') AS ?streamType).
-            |        BIND (IF(?streamType = 'stdout', '> ', IF(?streamType = 'stderr', '2> ', '')) AS ?streamOperator).
-            |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix) .
-            |        BIND (CONCAT(?prefix, ?streamOperator, STR(?value)) AS ?commandParameter) .
-            |      } UNION { # parameters
-            |        GRAPH <${GraphClass.Project.id(projectId)}> {
-            |          <$activityId> prov:qualifiedAssociation/prov:hadPlan ?planId.
-            |          ?planId renku:hasArguments ?parameter .
-            |          ?paramValue a renku:ParameterValue;
-            |                      schema:valueReference ?parameter .
-            |          ?paramValue schema:value ?value .
-            |          ?parameter renku:position ?position .
-            |          OPTIONAL { ?parameter renku:prefix ?maybePrefix }
-            |        }
-            |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix) .
-            |        BIND (CONCAT(?prefix, STR(?value)) AS ?commandParameter) .
-            |      }
-            |    }
-            |    GROUP BY ?position ?commandParameter
-            |    HAVING (COUNT(*) > 0)
-            |    ORDER BY ?position
-            |  }
-            |}
-            |GROUP BY ?command ?type ?location
-            |""".stripMargin
+        sparql"""|SELECT DISTINCT ?type (CONCAT(STR(?command), (GROUP_CONCAT(?commandParameter; separator=' '))) AS ?label) ?location
+                 |WHERE {
+                 |  {
+                 |    SELECT DISTINCT ?command ?type ?location
+                 |    WHERE {
+                 |      GRAPH ${GraphClass.Project.id(projectId)} {
+                 |        $activityId prov:qualifiedAssociation/prov:hadPlan ?planId;
+                 |                                      a ?type.
+                 |        OPTIONAL { ?planId renku:command ?maybeCommand. }
+                 |      }
+                 |      BIND (IF(bound(?maybeCommand), CONCAT(STR(?maybeCommand), STR(' ')), '') AS ?command).
+                 |      BIND ($activityId AS ?location)
+                 |    }
+                 |  } {
+                 |    SELECT ?position ?commandParameter
+                 |     WHERE {
+                 |      { # inputs
+                 |        GRAPH ${GraphClass.Project.id(projectId)} {
+                 |          $activityId prov:qualifiedAssociation/prov:hadPlan ?planId.
+                 |          ?planId renku:hasInputs ?input .
+                 |          ?paramValue a renku:ParameterValue ;
+                 |                      schema:valueReference ?input .
+                 |          ?paramValue schema:value ?value .
+                 |          OPTIONAL { ?input renku:mappedTo/renku:streamType ?maybeStreamType. }
+                 |          ?input renku:position ?position.
+                 |          OPTIONAL { ?input renku:prefix ?maybePrefix }
+                 |        }
+                 |        BIND (IF(bound(?maybeStreamType), ?maybeStreamType, '') AS ?streamType).
+                 |        BIND (IF(?streamType = 'stdin', '< ', '') AS ?streamOperator).
+                 |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix).
+                 |        BIND (CONCAT(?prefix, ?streamOperator, STR(?value)) AS ?commandParameter) .
+                 |      } UNION { # outputs
+                 |        GRAPH ${GraphClass.Project.id(projectId)} {
+                 |          $activityId prov:qualifiedAssociation/prov:hadPlan ?planId.
+                 |          ?planId renku:hasOutputs ?output .
+                 |          ?paramValue a renku:ParameterValue ;
+                 |                      schema:valueReference ?output .
+                 |          ?paramValue schema:value ?value .
+                 |          ?output renku:position ?position.
+                 |          OPTIONAL { ?output renku:mappedTo/renku:streamType ?maybeStreamType. }
+                 |          OPTIONAL { ?output renku:prefix ?maybePrefix }
+                 |        }
+                 |        BIND (IF(bound(?maybeStreamType), ?maybeStreamType, '') AS ?streamType).
+                 |        BIND (IF(?streamType = 'stdout', '> ', IF(?streamType = 'stderr', '2> ', '')) AS ?streamOperator).
+                 |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix) .
+                 |        BIND (CONCAT(?prefix, ?streamOperator, STR(?value)) AS ?commandParameter) .
+                 |      } UNION { # parameters
+                 |        GRAPH ${GraphClass.Project.id(projectId)} {
+                 |          $activityId prov:qualifiedAssociation/prov:hadPlan ?planId.
+                 |          ?planId renku:hasArguments ?parameter .
+                 |          ?paramValue a renku:ParameterValue;
+                 |                      schema:valueReference ?parameter .
+                 |          ?paramValue schema:value ?value .
+                 |          ?parameter renku:position ?position .
+                 |          OPTIONAL { ?parameter renku:prefix ?maybePrefix }
+                 |        }
+                 |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix) .
+                 |        BIND (CONCAT(?prefix, STR(?value)) AS ?commandParameter) .
+                 |      }
+                 |    }
+                 |    GROUP BY ?position ?commandParameter
+                 |    HAVING (COUNT(*) > 0)
+                 |    ORDER BY ?position
+                 |  }
+                 |}
+                 |GROUP BY ?command ?type ?location
+                 |""".stripMargin
       )
   }
 }

--- a/knowledge-graph/src/main/scala/io/renku/knowledgegraph/projects/files/lineage/NodeDetailsFinder.scala
+++ b/knowledge-graph/src/main/scala/io/renku/knowledgegraph/projects/files/lineage/NodeDetailsFinder.scala
@@ -161,12 +161,12 @@ private object NodeDetailsFinder {
                  |                                      a ?type.
                  |        OPTIONAL { ?planId renku:command ?maybeCommand. }
                  |      }
-                 |      BIND (IF(bound(?maybeCommand), CONCAT(STR(?maybeCommand), STR(' ')), '') AS ?command).
+                 |      BIND (IF(BOUND(?maybeCommand), CONCAT(STR(?maybeCommand), STR(' ')), '') AS ?command).
                  |      BIND ($activityId AS ?location)
                  |    }
                  |  } {
-                 |    SELECT ?position ?commandParameter
-                 |     WHERE {
+                 |    SELECT ?commandParameter
+                 |    WHERE {
                  |      { # inputs
                  |        GRAPH ${GraphClass.Project.id(projectId)} {
                  |          $activityId prov:qualifiedAssociation/prov:hadPlan ?planId.
@@ -174,14 +174,14 @@ private object NodeDetailsFinder {
                  |          ?paramValue a renku:ParameterValue ;
                  |                      schema:valueReference ?input .
                  |          ?paramValue schema:value ?value .
-                 |          OPTIONAL { ?input renku:mappedTo/renku:streamType ?maybeStreamType. }
-                 |          ?input renku:position ?position.
+                 |          OPTIONAL { ?input renku:mappedTo/renku:streamType ?maybeStreamType }
+                 |          OPTIONAL { ?input renku:position ?maybePosition }
                  |          OPTIONAL { ?input renku:prefix ?maybePrefix }
                  |        }
-                 |        BIND (IF(bound(?maybeStreamType), ?maybeStreamType, '') AS ?streamType).
+                 |        BIND (IF(BOUND(?maybeStreamType), ?maybeStreamType, '') AS ?streamType).
                  |        BIND (IF(?streamType = 'stdin', '< ', '') AS ?streamOperator).
-                 |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix).
-                 |        BIND (CONCAT(?prefix, ?streamOperator, STR(?value)) AS ?commandParameter) .
+                 |        BIND (IF(BOUND(?maybePrefix), STR(?maybePrefix), '') AS ?prefix).
+                 |        BIND (IF(BOUND(?maybePosition), CONCAT(?prefix, ?streamOperator, STR(?value)), '') AS ?commandParameter) .
                  |      } UNION { # outputs
                  |        GRAPH ${GraphClass.Project.id(projectId)} {
                  |          $activityId prov:qualifiedAssociation/prov:hadPlan ?planId.
@@ -189,14 +189,14 @@ private object NodeDetailsFinder {
                  |          ?paramValue a renku:ParameterValue ;
                  |                      schema:valueReference ?output .
                  |          ?paramValue schema:value ?value .
-                 |          ?output renku:position ?position.
-                 |          OPTIONAL { ?output renku:mappedTo/renku:streamType ?maybeStreamType. }
+                 |          OPTIONAL { ?output renku:position ?maybePosition }
+                 |          OPTIONAL { ?output renku:mappedTo/renku:streamType ?maybeStreamType }
                  |          OPTIONAL { ?output renku:prefix ?maybePrefix }
                  |        }
-                 |        BIND (IF(bound(?maybeStreamType), ?maybeStreamType, '') AS ?streamType).
+                 |        BIND (IF(BOUND(?maybeStreamType), ?maybeStreamType, '') AS ?streamType).
                  |        BIND (IF(?streamType = 'stdout', '> ', IF(?streamType = 'stderr', '2> ', '')) AS ?streamOperator).
-                 |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix) .
-                 |        BIND (CONCAT(?prefix, ?streamOperator, STR(?value)) AS ?commandParameter) .
+                 |        BIND (IF(BOUND(?maybePrefix), STR(?maybePrefix), '') AS ?prefix) .
+                 |        BIND (IF(BOUND(?maybePosition), CONCAT(?prefix, ?streamOperator, STR(?value)), '') AS ?commandParameter) .
                  |      } UNION { # parameters
                  |        GRAPH ${GraphClass.Project.id(projectId)} {
                  |          $activityId prov:qualifiedAssociation/prov:hadPlan ?planId.
@@ -204,16 +204,16 @@ private object NodeDetailsFinder {
                  |          ?paramValue a renku:ParameterValue;
                  |                      schema:valueReference ?parameter .
                  |          ?paramValue schema:value ?value .
-                 |          ?parameter renku:position ?position .
+                 |          OPTIONAL { ?parameter renku:position ?maybePosition }
                  |          OPTIONAL { ?parameter renku:prefix ?maybePrefix }
                  |        }
-                 |        BIND (IF(bound(?maybePrefix), STR(?maybePrefix), '') AS ?prefix) .
-                 |        BIND (CONCAT(?prefix, STR(?value)) AS ?commandParameter) .
+                 |        BIND (IF(BOUND(?maybePrefix), STR(?maybePrefix), '') AS ?prefix) .
+                 |        BIND (IF(BOUND(?maybePosition), CONCAT(?prefix, STR(?value)), '') AS ?commandParameter) .
                  |      }
                  |    }
-                 |    GROUP BY ?position ?commandParameter
+                 |    GROUP BY ?maybePosition ?commandParameter
                  |    HAVING (COUNT(*) > 0)
-                 |    ORDER BY ?position
+                 |    ORDER BY ?maybePosition
                  |  }
                  |}
                  |GROUP BY ?command ?type ?location

--- a/knowledge-graph/src/test/scala/io/renku/knowledgegraph/projects/files/lineage/NodeDetailsFinderSpec.scala
+++ b/knowledge-graph/src/test/scala/io/renku/knowledgegraph/projects/files/lineage/NodeDetailsFinderSpec.scala
@@ -252,7 +252,7 @@ class NodeDetailsFinderSpec
       }
 
       exception            shouldBe an[IllegalArgumentException]
-      exception.getMessage shouldBe s"No plan with $missingPlan"
+      exception.getMessage shouldBe s"No activity with $missingPlan"
     }
   }
 

--- a/knowledge-graph/src/test/scala/io/renku/knowledgegraph/projects/files/lineage/NodeDetailsFinderSpec.scala
+++ b/knowledge-graph/src/test/scala/io/renku/knowledgegraph/projects/files/lineage/NodeDetailsFinderSpec.scala
@@ -21,11 +21,12 @@ package io.renku.knowledgegraph.projects.files.lineage
 import LineageGenerators._
 import cats.effect.IO
 import cats.syntax.all._
-import io.renku.generators.Generators.Implicits._
 import io.renku.generators.Generators._
+import io.renku.generators.Generators.Implicits._
 import io.renku.graph.model.GraphModelGenerators.projectPaths
-import io.renku.graph.model.testentities.StepPlanCommandParameter._
+import io.renku.graph.model.parameterValues.ValueOverride
 import io.renku.graph.model.testentities._
+import io.renku.graph.model.testentities.StepPlanCommandParameter._
 import io.renku.interpreters.TestLogger
 import io.renku.jsonld.syntax._
 import io.renku.logging.TestSparqlQueryTimeRecorder
@@ -165,7 +166,46 @@ class NodeDetailsFinderSpec
         .unsafeRunSync() shouldBe Set(NodeDef(activity1).toNode, NodeDef(activity2).toNode)
     }
 
+    "find details of a Plan with implicit command parameters " +
+      "- no implicit params, inputs and outputs to be included in the node's label" in new TestCase {
+
+        val parameter = commandParameterDefaultValueGen.generateOne
+        val input     = entityLocations.generateOne
+        val output    = entityLocations.generateOne
+
+        val project = anyRenkuProjectEntities
+          .addActivity(project =>
+            executionPlanners(
+              stepPlanEntities(
+                CommandParameter.implicitFrom(parameter),
+                CommandInput.implicitFromLocation(input),
+                CommandOutput.implicitFromLocation(output)
+              ),
+              project
+            ).map(
+              _.planParameterValues(parameter -> ValueOverride(parameter.value))
+                .planInputParameterValuesFromChecksum(input -> entityChecksums.generateOne)
+                .buildProvenanceUnsafe()
+            )
+          )
+          .generateOne
+
+        upload(to = projectsDataset, project)
+
+        val activity :: Nil = project.activities
+
+        nodeDetailsFinder
+          .findDetails(
+            Set(
+              ExecutionInfo(activity.asEntityId.show, activity.startTime.value)
+            ),
+            project.path
+          )
+          .unsafeRunSync() shouldBe Set(NodeDef(activity).toNode)
+      }
+
     "find details of a Plan with mapped command parameters" in new TestCase {
+
       val input +: output +: errOutput +: Nil =
         entityLocations.generateNonEmptyList(min = 3, max = 3).toList
 

--- a/renku-model-tiny-types/src/main/scala/io/renku/graph/model/entityModel.scala
+++ b/renku-model-tiny-types/src/main/scala/io/renku/graph/model/entityModel.scala
@@ -19,6 +19,7 @@
 package io.renku.graph.model
 
 import cats.syntax.all._
+import cats.Show
 import io.renku.graph.model.entityModel.Location.FileOrFolder.from
 import io.renku.graph.model.views.{EntityIdJsonLDOps, TinyTypeJsonLDOps}
 import io.renku.jsonld.JsonLDDecoder.decodeString
@@ -80,6 +81,11 @@ object entityModel {
     }
 
     implicit lazy val jsonLDEncoder: JsonLDEncoder[Location] = encodeString.contramap(_.value)
+
+    implicit val show: Show[Location] = Show.show {
+      case Location.File(v)   => v
+      case Location.Folder(v) => v
+    }
   }
 
   final class Checksum private (val value: String) extends AnyVal with StringTinyType

--- a/renku-model/src/test/scala/io/renku/graph/model/testentities/LineageExemplarData.scala
+++ b/renku-model/src/test/scala/io/renku/graph/model/testentities/LineageExemplarData.scala
@@ -233,6 +233,7 @@ object NodeDef {
       .sortBy(_._2)
       .map(_._1.show)
       .mkString(start = commandComponent, sep = " ", end = "")
+      .trim
   }
 
   private implicit def parameterValueShow[P <: ParameterValue]: Show[P] = Show.show {

--- a/renku-model/src/test/scala/io/renku/graph/model/testentities/StepPlanCommandParameter.scala
+++ b/renku-model/src/test/scala/io/renku/graph/model/testentities/StepPlanCommandParameter.scala
@@ -21,11 +21,11 @@ package io.renku.graph.model.testentities
 import cats.syntax.all._
 import eu.timepit.refined.auto._
 import io.renku.cli.model.{CliCommandInput, CliCommandOutput}
+import io.renku.graph.model._
 import io.renku.graph.model.cli.CliConverters
-import io.renku.graph.model.commandParameters.IOStream.{StdErr, StdIn, StdOut}
 import io.renku.graph.model.commandParameters._
+import io.renku.graph.model.commandParameters.IOStream.{StdErr, StdIn, StdOut}
 import io.renku.graph.model.entityModel.Location
-import io.renku.graph.model.{RenkuUrl, commandParameters, entities, plans}
 import io.renku.jsonld._
 import io.renku.jsonld.syntax._
 
@@ -71,6 +71,17 @@ object StepPlanCommandParameter {
                                    maybePrefix = None,
                                    defaultValue = value,
                                    plan.id
+          )
+
+    def implicitFrom(value: ParameterDefaultValue): Position => Plan => ImplicitCommandParameter =
+      _ =>
+        plan =>
+          ImplicitCommandParameter(
+            Name(value.show),
+            maybeDescription = None,
+            maybePrefix = None,
+            value,
+            plan.id
           )
 
     implicit def toEntitiesCommandParameter(implicit
@@ -119,6 +130,17 @@ object StepPlanCommandParameter {
 
     def streamedFromLocation(defaultValue: Location): Position => Plan => CommandInput =
       streamedFrom(InputDefaultValue(defaultValue))
+
+    def implicitFromLocation(location: Location): Position => Plan => ImplicitCommandInput =
+      _ =>
+        plan =>
+          ImplicitCommandInput(
+            Name(location.show),
+            maybePrefix = None,
+            InputDefaultValue(location),
+            maybeEncodingFormat = None,
+            plan.id
+          )
 
     def from(defaultValue: InputDefaultValue): Position => Plan => LocationCommandInput =
       position =>
@@ -248,6 +270,18 @@ object StepPlanCommandParameter {
 
     def streamedFromLocation(defaultValue: Location, stream: IOStream.Out): Position => Plan => MappedCommandOutput =
       streamedFrom(OutputDefaultValue(defaultValue), stream)
+
+    def implicitFromLocation(location: Location): Position => Plan => ImplicitCommandOutput =
+      _ =>
+        plan =>
+          ImplicitCommandOutput(
+            Name(location.show),
+            maybePrefix = None,
+            OutputDefaultValue(location),
+            FolderCreation.no,
+            maybeEncodingFormat = None,
+            plan.id
+          )
 
     def stdOut(implicit renkuUrl: RenkuUrl): IOStream.StdOut =
       IOStream.StdOut(IOStream.ResourceId((renkuUrl / "iostreams" / StdOut.name).value))

